qualtran/bloqs/chemistry/prepare_mps/decompose_gate_hr.py

```diff
@@ -0,0 +1,240 @@
+    $$
+
+    For a detailed description of how to use this bloq refer to its tutorial.
+
+    Args:
+        phase_bitsize: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        gate_cols: tuple that contains, for each entry, a pair where the first element is the
+            index of the column and the second a tuple that contains the complex values of each
+            column. For example, for the gate [[a,b,],[c,d]] it would be ((0,(a,c)), (1(b,d))).
+        uncompute: wether to implement U or U^t. Either case the gate_cols to be provided are those
+            of U.
+        internal_phase_grad: a phase gradient state is needed for the decomposition. It can be
+            either be provided externally if this attribute is set to False or internally otherwise.
+        internal_refl_ancilla: this bloq also uses a reflection ancilla that can be provided
+            externally if this attribute is set to false or internally otherwise. In the case of the
+            internal use it will be projected into |0>, and all the data that was contained into the
+            |1> projection will be lost.
+
+    References:
+        [Synthesis of unitaries with Clifford+T circuits]
+        (https://arxiv.org/pdf/1306.3200.pdf).
+            Kliuchnikov 2013.
+    """
+    phase_bitsize: int  # number of ancilla qubits used to encode the state preparation's rotations
+    gate_cols: Tuple[
+        int, Tuple[complex, ...]
+    ]  # tuple with the columns of the gate that are specified
+    uncompute: bool = False
+    internal_phase_grad: bool = False
+    internal_refl_ancilla: bool = True
+
+    def __attrs_post_init__(self):
+        # at least one column has to be specified
+        assert len(self.gate_cols) > 0
+        # there can't be more columns that rows
+        assert len(self.gate_cols) <= len(self.gate_cols[0][1])
+        # all cols must be the same length and a power of two
+        lengths = set([len(c[1]) for c in self.gate_cols])
+        assert len(lengths) == 1
+        assert list(lengths)[0] == 2**self.gate_bitsize
+
+    @property
+    def signature(self):
+        return Signature.build(
+            refl_ancilla=(not self.internal_refl_ancilla),
+            gate_input=self.gate_bitsize,
+            phase_grad=(not self.internal_phase_grad) * self.phase_bitsize,
+        )
+
+    @property
+    def gate_bitsize(self):
+        return (len(self.gate_cols[0][1]) - 1).bit_length()
+
+    def build_composite_bloq(
+        self, bb: BloqBuilder, *, gate_input: SoquetT, **soqs: SoquetT
+    ) -> Dict[str, SoquetT]:
+        """Extra soquets inside soqs are:
+        * phase_grad: a phase gradient state of size phase_bitsize if internal_phase_gradient
+                    is set to False
+        * refl_ancilla: a clean qubit in |0> if internal_refl_ancilla is set to False
+        """
```

Comment:
this can probably just be line comments `#` instead of a docstring. If you want this to be a docstring, it should follow the docstring format. 

By "extra soquets" you mean "optional soquets"?

Reply:
Not really optional, but required in some situations. For example, if an user sets the `internal_phase_grad` flag to `False` then that register has to be provided, but it is not optional in that case.

Reply:
optional in a cosmic sense where you can set the flag to true